theseus/geometry/so3.py

```diff
@@ -0,0 +1,231 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional, Union, cast
+
+import torch
+
+import theseus.constants
+
+from .lie_group import LieGroup
+from .point_types import Point3
+
+
+class SO3(LieGroup):
+    SO3_EPS = 5e-7
+
+    def __init__(
+        self,
+        quaternion: Optional[torch.Tensor] = None,
+        data: Optional[torch.Tensor] = None,
+        name: Optional[str] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        if quaternion is not None and data is not None:
+            raise ValueError("Please provide only one of quaternion or data.")
+        if quaternion is not None:
+            dtype = quaternion.dtype
+        if data is not None:
+            self._SO3_matrix_check(data)
+        super().__init__(data=data, name=name, dtype=dtype)
+        if quaternion is not None:
+            if quaternion.ndim == 1:
+                quaternion = quaternion.unsqueeze(0)
+            self.update_from_unit_quaternion(quaternion)
+
+    @staticmethod
+    def _init_data() -> torch.Tensor:  # type: ignore
+        return torch.eye(3, 3).view(1, 3, 3)
+
+    def update_from_unit_quaternion(self, quaternion: torch.Tensor):
+        self.update(self.unit_quaternion_to_matrix(quaternion))
+
+    def dof(self) -> int:
+        return 3
+
+    def __repr__(self) -> str:
+        return f"SO3(data={self.data}, name={self.name})"
+
+    def __str__(self) -> str:
+        with torch.no_grad():
+            return f"SO3(matrix={self.data}), name={self.name})"
+
+    def _adjoint_impl(self) -> torch.Tensor:
+        return self.data.clone()
+
+    @staticmethod
+    def _SO3_matrix_check(matrix: torch.Tensor):
+        if matrix.ndim != 3 or matrix.shape[1:] != (3, 3):
+            raise ValueError("3D rotations can only be 3x3 matrices.")
+        _check = (
+            torch.matmul(matrix, matrix.transpose(1, 2))
+            - torch.eye(3, 3, dtype=matrix.dtype, device=matrix.device)
+        ).abs().max().item() < SO3.SO3_EPS
+        _check &= (torch.linalg.det(matrix) - 1).abs().max().item() < SO3.SO3_EPS
+
+        if not _check:
+            raise ValueError("Not valid 3D rotations.")
+
+    @staticmethod
+    def _unit_quaternion_check(quaternion: torch.Tensor):
+        if quaternion.ndim != 2 or quaternion.shape[1] != 4:
+            raise ValueError("Quaternions can only be 4-D vectors.")
+
+        if (torch.linalg.norm(quaternion, dim=1) - 1).abs().max().item() >= SO3.SO3_EPS:
+            raise ValueError("Not unit quaternions.")
+
+    @staticmethod
+    def exp_map(tangent_vector: torch.Tensor) -> LieGroup:
```

Comment:
To confirm the exp and log maps are novel so we don't have anything to reference? For now it would be useful to add some high level comments explaining the methods.

Reply:
These exp and log maps implemented here are common. The novel retraction maps have not been implemented.